enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java

```diff
@@ -0,0 +1,365 @@
+                for ( Entity entity : entities )
+                {
+                    long entityId = entity.getId();
+                    Map<String,Object> properties = entity.getProperties( indexedPropertyKeys );
+                    if ( !properties.isEmpty() )
+                    {
+                        documents.add( documentForPopulation( entityId, properties ) );
+                    }
+
+                    if ( documents.size() > POPULATING_BATCH_SIZE )
+                    {
+                        indexWriter.addDocuments( documents.size(), documents );
+                        documents.clear();
+                    }
+                }
+            }
+            indexWriter.addDocuments( documents.size(), documents );
+            return Pair.of( index, completedLatch );
+        };
+
+        enqueueUpdate( population );
+        return completedLatch;
+    }
+
+    private void enqueueUpdate( FulltextIndexUpdate update ) throws IOException
+    {
+        try
+        {
+            workQueue.put( update );
+        }
+        catch ( InterruptedException e )
+        {
+            throw new IOException( "Fulltext index update failed.", e );
+        }
+    }
+
+    void start()
+    {
+        if ( workerThread != null )
+        {
+            throw new IllegalStateException( workerThread.getName() + " already started." );
+        }
+        workerThread = new ApplierThread( workQueue, log, availabilityGuard );
+        workerThread.start();
+    }
+
+    void stop()
+    {
+        boolean enqueued;
+        do
+        {
+            enqueued = workQueue.offer( STOP_SIGNAL );
+        }
+        while ( !enqueued );
+
+        try
+        {
+            workerThread.join();
+            workerThread = null;
+        }
+        catch ( InterruptedException e )
+        {
+            log.error( "Interrupted before " + workerThread.getName() + " could shut down.", e );
+        }
+    }
+
+    private interface FulltextIndexUpdate
+    {
+        Pair<WritableFulltext,BinaryLatch> applyUpdateAndReturnIndex() throws IOException;
+    }
+
+    private static class ApplierThread extends Thread
+    {
+        private LinkedBlockingQueue<FulltextIndexUpdate> workQueue;
+        private final Log log;
+        private final AvailabilityGuard availabilityGuard;
+
+        ApplierThread( LinkedBlockingQueue<FulltextIndexUpdate> workQueue, Log log,
+                       AvailabilityGuard availabilityGuard )
+        {
+            super( "Fulltext Index Add-On Applier Thread" );
+            this.workQueue = workQueue;
+            this.log = log;
+            this.availabilityGuard = availabilityGuard;
+            setDaemon( true );
+        }
+
+        @Override
+        public void run()
+        {
+            waitForDatabaseToBeAvailable();
+            Set<WritableFulltext> refreshableSet = Collections.newSetFromMap( new IdentityHashMap<>() );
```

Comment:
why set can't be simple HashSet, for example? why do we need identity hash map in here?

Reply:
still open question here